lib/jwt/x509/validator.rb

```diff
@@ -0,0 +1,32 @@
+module JWT
+  module X509
+    class Validator
+      # @param [Hash] opts x509 options for finding key
+      # @option opts [String] x5u: (not yet supported)
+      # @option opts [Array<String>] x5c: x509 cert chain, where first is the cert used to sign
+      # @option opts [String] x5t: (not yet supported)
+      def initialize(opts)
+        x5c = opts[:x5c]
+
+        unless x5c.nil?
+          signing_der = ::Base64.decode64 x5c.first
+          @signing_cert = OpenSSL::X509::Certificate.new signing_der
+          len = x5c.length
+          @cert_chain = x5c[1...len].map do |b64der|
+            OpenSSL::X509::Certificate.new(::Base64.decode64(b64der))
+          end
+        end
+      end
+
+      def valid?
+        return true if @cert_chain.blank?
+
+        store = OpenSSL::X509::Store.new
+        @cert_chain.each do |cert|
+          store.add_cert cert
+        end
+        store.verify @signing_cert
+      end
```

Comment:
This block of code has security problems. From [RFC 5280 section 3.2](https://tools.ietf.org/html/rfc5280#section-3.2):

> In general, a chain of multiple certificates may be needed, comprising a certificate of the public key owner (the end entity) signed by one CA, and zero or more additional certificates of CAs signed by other CAs.  Such chains, called certification paths, are required because a public key user is only initialized with a limited number of assured CA public keys.

More concretely:
- Chain validation can not be skipped: this would allow an attacker to just strip the chain and inject their own certificate and use it to sign, bypassing any security check. Even a single certificate must chain correctly to a trust anchor.
- The trust anchor should be configurable by the consuming app. It may be tempting to default to `OpenSSL::X509::Store#set_defaults_path` but instead of allowing all the certificate authorities to sign, the user should specify this themselves and preferably only limit to root certificate(s) that apply for their use case. E.g. when validating a JWT from Google, use only the cert(s) from https://pki.goog/
- [RFC 5280 section 3.3](https://tools.ietf.org/html/rfc5280#section-3.3) also mentions revocation checking, which is absent in this PR.

I've been implementing all of the above in the WebAuthn gem, you can find my code here: https://github.com/cedarcode/webauthn-ruby/pull/208 - feel free to use it, I'd appreciate a [mention as co-author](https://help.github.com/en/articles/creating-a-commit-with-multiple-authors) if you do 🙂

Reply:
@bdewater sorry for the late response here! Thank you for all the feedback here, definitely seems like I was missing some parts. And that webauthn PR looks impressive.

Unfortunately I am no longer working on this PR, as my project's requirements changed and I don't need to use `x5c` for JWTs anymore. Please feel free to take my work and continue it if you'd like!

Reply:
@bdewater would you be interested in taking over this PR, because it looks like a important feature. Newer versions of the IDP's are supporting the x5c  header and I think we should support it in this gem too. 

Let me know if you would be able to work on this one, else I can pick it up and resolve :)

Reply:
I'm currently on vacation but was planning to pick this up again after I'm back next week :)

Reply:
See https://github.com/jwt/ruby-jwt/pull/338